### ISSUE TITLE
`fromfile` should raise an exception if not passed a “low-level file”

### ISSUE THREAD

**Issue by gerritholl** [numpy/numpy#7713]
An [ancient ticket from 2007](https://github.com/numpy/numpy/issues/1103) notes that `fromfile` does not work with files opened with gzip.  Instead, it will interpret the literal gzip data.  This is due to the fundamental way `fromfile` works and the resolution of the bug was essentially wontfix.

Secondly, it should be made clear in the documentation that you cannot pass a `GzipFile`.  Some file-like objects trigger `io.UnsupportedOperation`.

It may save bugs and headaches if `fromfile` were to raise an Exception if it is passed a file-object that is not a low-level file.

**Comment by birm** [comment]
Do you think that fromfile should create a temporary low level file from a file object, if that even works?

What happens when you use shutil.copyfileobj() on these file objects and operate on the copy?

**Comment by gerritholl** [comment]
I don't know how one could create a temporary low level file from a file object.  My current workaround is to use `frombuffer` instead of `fromfile`.  I'm not quite sure what you mean or want to achieve with `copyfileobj`.  A low-level file would by definition map to a file on the file system (which may or may not correspond to a file on disk), wouldn't it?  In that case, the `copyfileobj` would have no benefit over reading the contents into memory, then using `frombuffer`.

If there is a straightforward and foolproof way for `fromfile` to tell that it is not operating on a low-level file, it could raise a `TypeError` or some other exception, while hinting that the user may wish to use `frombuffer` instead.

**Comment by birm** [comment]
Sorry, I think I'm not quite understanding, and may still not be, but isn't that more of an issue with gzipfile since it doesn't have an implemented method of io readinto, but inherits it in a way that produces undesirable results? In the cases where you get UnsupportedOperation, it isn't that numpy is raising it, the file object is.

**Comment by gerritholl** [comment]
I don't know how to tell whether this is an issue that can/should be fixed or raised in `numpy.fromfile` or in one of `GzipFile`s methods, and if so, which one, and how.

**Comment by birm** [comment]
I think it's an issue best suited to GzipFile, [so I put in a patch](http://bugs.python.org/issue27345).

**Comment by gerritholl** [comment]
According to your patch report, `GzipFile.readinto` works as expected in Python 3.5.  But `numpy.fromfile` does not.  Then where does the problem lie?

**Comment by birm** [comment]
I'm having some issues replicating your issue, can you help me out some?

**Comment by QuLogic** [comment]
Using a bit of data from matplotlib so I don't have to attach it here; feel free to use your own gzipped data:
```python
In [1]: from matplotlib import cbook
In [2]: import numpy as np

In [3]: dfile = cbook.get_sample_data('s1045.ima.gz')
In [4]: im1 = np.fromstring(dfile.read(), np.uint16)
In [5]: dfile.close()

In [6]: dfile = cbook.get_sample_data('s1045.ima.gz')
In [7]: im2 = np.fromfile(dfile, np.uint16)
In [8]: dfile.close()

In [9]: im1.shape
Out[9]: (65536,)

In [10]: im2.shape
Out[10]: (16614,)

In [11]: im1[:10]
Out[11]: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=uint16)

In [12]: im2[:10]
Out[12]: array([35615,  2056, 11674, 20542,   768, 12659, 13360, 11829, 28009,    97], dtype=uint16)
```
using Python 3.5.2 and NumPy 1.11.0. It works in that it doesn't _crash_, but it doesn't really _work_.

**Comment by jzwinck** [comment]
Is there any reason not to implement this so it actually works?  It's been ten years since the bug was initially reported, and it's still hurting people--for example https://stackoverflow.com/questions/47574607/python-gzip-module-doesnt-work-as-expected-on-ubyte-file/

It seems to me that it is possible to implement `np.fromfile()` so that it works with `GzipFile` (but not using `readinto()`).

**Comment by mhvk** [comment]
@jzwinck - the reason for not doing something in a project run by volunteers is usually lack of time or personal interest, not that a given issue isn't deserving...

In other words, it would be *great* if you or someone else stepped up and made a PR that allows `fromfile` to work properly, or at least do what is requested here, raise an error when it will fail and document work-arounds.

**Comment by mattip** [comment]
Closing. If [`np.fromtxt`](http://www.numpy.org/devdocs/reference/generated/numpy.loadtxt.html?highlight=loadtxt#numpy.loadtxt) does not work sufficiently well with compressed data, please open a new issue.

**Comment by QuLogic** [comment]
`np.fromfile` accepts data from a text or binary file; `np.fromtxt` is only for text.

**Comment by mattip** [comment]
Sorry, I meant `np.loadtxt` which accepts a filename

**Comment by QuLogic** [comment]
Again, that only accepts text files, `fromfile` is for binary data.

**Comment by jzwinck** [comment]
@mattip I think you may be missing the severity of the problem here.  This is a data corruption bug.  This ticket asks for an exception to be raised by `np.fromfile()` rather than loading corrupt data.  It has nothing to do with text files.  We don't have to implement the loading properly, but letting it load corrupt data with no complaint is bad.

**Comment by gerritholl** [comment]
> @jzwinck - the reason for not doing something in a project run by volunteers is usually lack of time or personal interest, not that a given issue isn't deserving...

@mhvk That is true, of course, but if an issue is labelled as "wontfix", I would interpret that as a declaration that the issue isn't desirable.  For issues that are desirable but that nobody has taken the time to fix, I would expect it to remain open, perhaps labelled "help wanted".

**Comment by gerritholl** [comment]
@mattip I was not trying to read compressed text data.  I was trying to read compressed binary data, and I was getting corrupt data, and it took me a while to figure out why, because I thought that `fromfile` would read from any file object.  It's fine if it can't, but it would have saved me time if I had gotten an exception or (at least) if it was documented to not work.  Currently, the `fromfile` documentation just says "Open file object or filename.".  Is a GzipFile a file object?  Some people, like me, may believe it is.

**Comment by gerritholl** [comment]
According to the Python glossary, a [file object](https://docs.python.org/3/glossary.html#term-file-object) is:

> An object exposing a file-oriented API (with methods such as read() or write()) to an underlying resource. Depending on the way it was created, a file object can mediate access to a real on-disk file or to another type of storage or communication device (for example standard input/output, in-memory buffers, sockets, pipes, etc.). File objects are also called file-like objects or streams.

That means that, according to the combined documentation of numpy and python, passing a GzipFile to `fromfile` should be expected to work correctly.  If it can't, that should be at least documented, and if not too expensive, verified runtime.

I disagree with the closing of this issue.

**Comment by eric-wieser** [comment]
I think I agree with the consensus here that we should probably reopen this. Since closed issues are hard to rediscover, I'm going to reopen it for now - we can always close it again if another maintainer wants to argue strongly for it.

Sometimes I wish we had fewer APIs for serialization / deserialization, especially for text formats...

**Comment by sidhant007** [comment]
Hey, I wish to take this issue and made some progress in reading the codebase around this portion.

In `numpy/core/src/multiarray/multiarraymodule.c` in the `array_fromfile` method it seems feasible to check if the file in hand is a gzip or not and accordingly throw an error.

However if the detection method consists of seeing the first 3 bytes and matching them with "\x1f\x8b\x08", then we cannot differentiate between gzip files and those binary files which also start with the same first 3 bytes (co-incidentally). 

@eric-wieser do you have any suggestions on how can the detection method look like?

**Comment by seberg** [comment]
I just stumbled on `io.RawIOBase`, I think giving a `UserWarning` when `not isinstance(filelike, io.RawIOBase)` may be reasonable (right now, I am not sure what the exact definition of `io.RawIOBase` is, but `issubclass(gzip.GzipFile, io.RawIOBase)` is False).
